xmtp_api_grpc/src/replication_client.rs

```diff
@@ -170,207 +160,166 @@ impl XmtpApiClient for ClientV4 {
     }
 
     #[tracing::instrument(level = "trace", skip_all)]
     async fn fetch_key_packages(
         &self,
         req: FetchKeyPackagesRequest,
     ) -> Result<FetchKeyPackagesResponse, Error> {
-        unimplemented!();
+        let topics = req
+            .installation_keys
+            .iter()
+            .map(|key| build_key_package_topic(key.as_slice()))
+            .collect();
+
+        let envelopes = self.query_v4_envelopes(topics).await?;
+        let key_packages: Result<Vec<_>, Error> = envelopes
+            .iter()
+            .map(|envelopes| {
+                // The last envelope should be the newest key package upload
+                let unsigned = envelopes.last().ok_or_else(|| {
+                    Error::new(ErrorKind::InternalError(InternalError::MissingPayloadError))
+                        .with("No envelopes found")
+                })?;
+
+                let client_env = extract_client_envelope(unsigned)?;
+
+                if let Some(Payload::UploadKeyPackage(upload_key_package)) = client_env.payload {
+                    let key_package = upload_key_package.key_package.ok_or_else(|| {
+                        Error::new(ErrorKind::InternalError(InternalError::MissingPayloadError))
+                            .with("Missing key package")
+                    })?;
+
+                    Ok(fetch_key_packages_response::KeyPackage {
+                        key_package_tls_serialized: key_package.key_package_tls_serialized,
+                    })
+                } else {
+                    Err(
+                        Error::new(ErrorKind::InternalError(InternalError::MissingPayloadError))
+                            .with("Payload is not a key package"),
+                    )
+                }
+            })
+            .collect();
+
+        Ok(FetchKeyPackagesResponse {
+            key_packages: key_packages?,
+        })
     }
 
     #[tracing::instrument(level = "trace", skip_all)]
     async fn send_group_messages(&self, req: SendGroupMessagesRequest) -> Result<(), Error> {
-        unimplemented!();
+        self.publish_envelopes_to_payer(req.messages).await
     }
 
     #[tracing::instrument(level = "trace", skip_all)]
     async fn send_welcome_messages(&self, req: SendWelcomeMessagesRequest) -> Result<(), Error> {
-        unimplemented!();
+        self.publish_envelopes_to_payer(req.messages).await
     }
 
     #[tracing::instrument(level = "trace", skip_all)]
     async fn query_group_messages(
         &self,
         req: QueryGroupMessagesRequest,
     ) -> Result<QueryGroupMessagesResponse, Error> {
-        unimplemented!();
+        let client = &mut self.client.clone();
+        let res = client
+            .query_envelopes(QueryEnvelopesRequest {
+                query: Some(EnvelopesQuery {
+                    topics: vec![build_group_message_topic(req.group_id.as_slice())],
+                    originator_node_ids: vec![],
+                    last_seen: None,
```

Comment:
As a note - the `PagingInfo` of the `QueryGroupMessagesRequest` has an `id_cursor` that we could potentially convert into vector clock format here as a short-term hack. Unsure if we get subtle bugs without this, as the client expects to query from where it left off in the queue. In the long-term we need to do a bigger refactor to properly support vector clocks though, just making a note in case you run into bugs in the meantime